### ISSUE TITLE
SFTP repo fails to create when using with newer SSH server and RSA keys - newer libssh2 needed

### ISSUE THREAD

**Issue by delavei** [pgbackrest/pgbackrest#2113]
Information about software:
1. pgBackRest version: 2.46

2. PostgreSQL version: 15.1

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
Docker containers using Debian 12.0 (bookworm), host running  Linux 5.15.0-75
SFTP server container is using OpenSSH_9.2p1 Debian-2, OpenSSL 3.0.9

4. Did you install pgBackRest from source or from a package? package

5. I attach the relevant output of my pgbackrest create-sanza command with debug log level.
DEBUG:     storage/sftp/storage::storageSftpNew: (path: {"/home/pgbackrest"}, host: {"database_backup_repo"}, port: 22, user: {"pgbackrest"}, timeout: 60000, keyPriv: {"/var/lib/postgresql/.ssh/id_rsa"}, hostKeyHashType: sha256, param.keyPub: {"/var/lib/postgresql/.ssh/id_rsa.pub"}, param.hostFingerprint: null, param.modeFile: 0640, param.modePath: 0750, param.write: false, param.pathExpressionFunction: (function *))
DEBUG:     common/io/socket/client::sckClientNew: (host: {"database_backup_repo"}, port: 22, timeoutConnect: 60000, timeoutSession: 60000)
DEBUG:     common/io/socket/client::sckClientNew: => {type: socket, driver: {host: database_backup_repo, port: 22, timeoutConnect: 60000, timeoutSession: 60000}}
DEBUG:     storage/sftp/storage::storageSftpNew: attempting public key authentication
DEBUG:     command/exit::exitSafe: (result: 0, error: true, signalType: 0)
ERROR: [101]: public key authentication failed: libssh2 error [-18]
       HINT: libssh2 compiled against non-openssl libraries requires --repo-sftp-private-key-file and --repo-sftp-public-key-file to be provided
       HINT: libssh2 versions before 1.9.0 expect a PEM format keypair, try ssh-keygen -m PEM -t rsa -P "" to generate the keypair
       --------------------------------------------------------------------
       If SUBMITTING AN ISSUE please provide the following information:
       
       version: 2.46
       command: stanza-create
       options: --exec-id=809-29d68ed2 --log-level-console=debug --pg1-path=/var/lib/pgsql/data --repo1-cipher-pass=<redacted> --repo2-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo2-cipher-type=aes-256-cbc --repo1-path=/var/lib/pgbackrest --repo2-path=/home/pgbackrest --repo2-sftp-host=database_backup_repo --repo2-sftp-host-key-hash-type=sha256 --repo2-sftp-host-user=pgbackrest --repo2-sftp-private-key-file=/var/lib/postgresql/.ssh/id_rsa --repo2-sftp-public-key-file=/var/lib/postgresql/.ssh/id_rsa.pub --repo2-type=sftp --stanza=finplay
       
       stack trace:
       storage/sftp/storage.c:storageSftpEvalLibSsh2Error:131:(test build required for parameters)
           ... function(s) omitted ...
       storage/sftp/storage.c:storageSftpNew:(path: {"/home/pgbackrest"}, host: {"database_backup_repo"}, port: 22, user: {"pgbackrest"}, timeout: 60000, keyPriv: {"/var/lib/postgresql/.ssh/id_rsa"}, hostKeyHashType: sha256, param.keyPub: {"/var/lib/postgresql/.ssh/id_rsa.pub"}, param.hostFingerprint: null, param.modeFile: 0640, param.modePath: 0750, param.write: false, param.pathExpressionFunction: (function *))
       storage/sftp/helper.c:storageSftpHelper:(repoIdx: 1, write: false, pathExpressionCallback: *void)
       command/stanza/create.c:cmdStanzaCreate:(storage: {type: posix, path: /var/lib/pgbackrest, write: false}, pgVersion: 150000, pgSystemId: 7194441522346983465, cipherType: aes-256-cbc)
       main.c:main:(debug log level required for parameters)
       --------------------------------------------------------------------

Output form debug log level of OpenSSH server:

| Connection from 10.0.1.4 port 59880 on 10.0.1.8 port 22 rdomain ""
| debug1: Local version string SSH-2.0-OpenSSH_9.2p1 Debian-2
| debug1: Remote protocol version 2.0, remote software version libssh2_1.10.0
| debug1: compat_banner: no match: libssh2_1.10.0
| debug1: permanently_set_uid: 101/65534 [preauth]
| debug1: list_hostkey_types: ssh-ed25519 [preauth]
| debug1: SSH2_MSG_KEXINIT sent [preauth]
| debug1: SSH2_MSG_KEXINIT received [preauth]
| debug1: kex: algorithm: curve25519-sha256 [preauth]
| debug1: kex: host key algorithm: ssh-ed25519 [preauth]
| debug1: kex: client->server cipher: aes128-ctr MAC: hmac-sha2-256 compression: none [preauth]
| debug1: kex: server->client cipher: aes128-ctr MAC: hmac-sha2-256 compression: none [preauth]
| debug1: expecting SSH2_MSG_KEX_ECDH_INIT [preauth]
| debug1: SSH2_MSG_KEX_ECDH_INIT received [preauth]
| debug1: rekey out after 4294967296 blocks [preauth]
| debug1: SSH2_MSG_NEWKEYS sent [preauth]
| debug1: expecting SSH2_MSG_NEWKEYS [preauth]
| debug1: SSH2_MSG_NEWKEYS received [preauth]
| debug1: rekey in after 4294967296 blocks [preauth]
| debug1: KEX done [preauth]
| debug1: userauth-request for user pgbackrest service ssh-connection method publickey [preauth]
| debug1: attempt 0 failures 0 [preauth]
| debug1: user pgbackrest matched 'User pgbackrest' at line 126
| debug1: authentication methods list 0: publickey
| debug1: authentication methods list 0: publickey [preauth]
| debug1: PAM: initializing for "pgbackrest"
| debug1: PAM: setting PAM_RHOST to "10.0.1.4"
| debug1: PAM: setting PAM_TTY to "ssh"
| userauth_pubkey: signature algorithm ssh-rsa not in PubkeyAcceptedAlgorithms [preauth]
| Connection closed by authenticating user pgbackrest 10.0.1.4 port 59880 [preauth]

7. Describe the issue:
When I try to initialize stanza using SFTP repo with afforementioned software used, pgbackrest fails. I identified the issue being that libssh2 uses ssh-rsa signature algorithm as default for RSA keys. However, this method was announced to be deprecated from next version in OpenSSH 8.7 (https://www.openssh.com/txt/release-8.7).
Whole authentication process would perhaps work by adding HostKeyAlgorithms=+ssh-rsa into sshd_config. However if changing of server's sshd_config is not possible (or wanted), it would be useful to have this ability in pgBackRest. 
This seems to be libssh2 issue (https://github.com/libssh2/libssh2/issues/536). At the end of the thread, there is notice, that support for this was added in libssh2 1.11.0, however according to "apt cache show pgbackrest" only libssh2-1 >= 1.5.0 is needed for pgBackRest 2.46. It would be useful to use newer version of libssh2 when building pgBackRest.

**Comment by jreidthompson** [comment]
I believe that this issue is caused by current distributions/openssh configurations excluding RSA type keys by default. Note the log message " userauth_pubkey: signature algorithm ssh-rsa not in PubkeyAcceptedAlgorithms [preauth]" .  I believe that you will either have to configure usage of a more secure algorithm or configure your ssh server to accept rsa.

**Comment by delavei** [comment]
Yes, just as you suggested, I resolved this issue by using different algorithm keypair (I forgot to mention it). I was just suggesting that this may be useful for some people forced to use older keys with newer SSH servers, although I personally can't come up with scenario, where this may happen (but I have very limited knowledge of different deployment scenarios than mine).

What I meant by using newer version of libssh2 when building pgBackRest was to use newer package as a dependency. I apologize for this wrong wording, you are of course right, that package build is controlled by the distribution. I realized this just a minute after I posted whole issued, that this can be resolved by manually installing newer libssh2.

I guess since two possible solutions were mentioned (different keypair, newer libssh2), both of which are in addition not even directly connected to pgBackRest, this may be closed.
